Support auxiliary neighbor variables for FE IKs

GiudGiud opened the pull request:
Refs #28977

the issue also exists for FVIks but they have logic there for multi-system that prevent any error / seg fault. 
The fact that we hit this problem probably means we are missing multi-system logic for regular IKs

GiudGiud committed:
Misc doco fixups

GiudGiud committed:
Forbid using auxiliary variables for the neighbor variable in IKs
closes #28977

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2524665/), step Docs: sync website on ced9d7e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28978/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2524712/), step Generate coverage on ced9d7e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b8db75</th>
      <th colspan=4>#28978 ced9d7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.08%</td>
      <td>85.08%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>106569</td>
      <td>106571</td>
      <td>+2</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18694</td>
      <td>18694</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28978/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28978/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28978/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad reviewed (changes requested):
I don't know if I'm ready to say we should ban this. I'd want to see if whether we gracefully early returned whether the user that prompted this has a good/valid formulation

GiudGiud committed:
Allow auxiliary variables in interface kernels
- check for different systems being used for var and neighbor_var
- only contribute to Jacobian for neighbor rows when neighbor_var is in the same system

GiudGiud committed:
Add testing for interface kernels support for auxiliary variables
with and without residual and Jac together

GiudGiud committed:
Add testing for FV Iks support for aux neighbor variables

GiudGiud commented:
This should be ready. This also allows multi-system FE Iks most likely (just need to have 2 IKs instead of 1)

lindsayad commented:
lot of failures 😦

GiudGiud commented:
it was passing then I did something bad

GiudGiud commented:
I think I simply pushed without pulling first. So should be good.

lindsayad commented on framework/src/interfacekernels/InterfaceKernel.C:
```diff
@@ -216,7 +220,9 @@ InterfaceKernelTempl<T>::computeResidual()
 
   // Compute the residual for the neighbor
-  computeElemNeighResidual(Moose::Neighbor);
+  // This also prevents computing a residual if the neighbor variable is auxiliary
+  if (_same_system)
+    computeElemNeighResidual(Moose::Neighbor);
```
So if the neighbor variable was from a nonlinear system that is different than the element variable, would the neighbor residual ever get computed?

lindsayad replied:
does your code work if the aux variable is the element var vs. the neighbor var?

GiudGiud replied:
No there s no way. You can't set the variable parameter to an aux variable, you ll get a "did you mean to use ..." from the MooseVariableInterface

lindsayad replied:
Hmm, I did not think through all the repercussions before sending you down this goose chase. My inclination at this point is to ban mixed systems since it seems odd to support mixed nl-aux only "by luck", e.g. the user happens to assign the aux variable to the neighbor side. I suppose if it was well documented in the interface kernel that you can use an aux variable, but only on the neighbor side, then we could move ahead. What do you think?

GiudGiud replied:
As you want. I'd merge this since it fixes the seg fault the user got without breaking anything. I can also work on making multi system work. I think I singled out the places to add code to already at this point.

lindsayad replied:
We can merge, but can you add something in `framework/doc/content/syntax/InterfaceKernels/index.md`?

GiudGiud replied:
done

GiudGiud committed:
Add note in documentation about the current state of support for multiple systems

GiudGiud commented:
Thanks for the review!
and thanks to the TMAP community user who reported this
